### ISSUE TITLE
Multiple Linux flavours for one compiler (devtoolset)

### ISSUE THREAD

**Issue by theirix** [conan-io/conan#707]
Hello.
It is rather a discussion issue than an bug.

A few words about devtoolset. Let we have two different versions of Linux - Redhat 6 (el6) and Redhat 7 (el7). Redhat developers often use a special [devtoolset](https://www.softwarecollections.org/en/scls/rhscl/devtoolset-4/) that allows to use a modern compiler (gcc 5.2) for el6 and el7. Devtoolset uses static libc/libstdc++ linking with bundled libraries so a resulting binary can run on a target system with much older libc/libstdc++. Of course, binaries compiled with devtoolset at el6 can be run only at el6 and so on. Linking with el7-based binary probably will fail because it targets different LIBC:
`./bin7: /lib64/libc.so.6: version `GLIBC_2.14' not found (required by ./bin7)`

When we are going to use conan libraries, they are selected only based on compiler (gcc 5.2) and OS type (Linux). We are going to rebuild all packages at our organization's CI. If libraries are being rebuilt for these two platforms it is not possible to distinguish el6 and el6 in a conan cache. And this is a problem.

Possible variants to solve a problem:
1. Copy recipe. I tried to `conan copy` to assign a different `user/channel` pair (i.e. myorg/el6) but I think the same recipe cannot contain two different packages. Is this true?
2. Adjust version. Prepare copies of recipes and specify different versions (such as `libxml2/2.9.3el6@myorg/stable`). It is not very convenient because requires to copy and track all recipes or automatically `sed` them.
3. Add additional option axis...
4. Change "os" setting to "Linux_el6"...

What is the best way to organize packages for this usecase?

**Comment by lasote** [comment]
Hi @theirix. It's an interesting discussion.
We already know about this problems with libc, actually we discussed about the introduction of a new setting for that. But we think it is not common to install newer compilers in an older distribution, so by default we didn't wanted to introduce a new setting.
Users usually don't know what libc version they have installed. And the introduction of a new setting would require detect the installed libc and/or ask users to set the right libc version.

But Conan is flexible enough to manage it! My recommendation is to add a new setting "libc" to your settings file `.conan/settings.yml` and create your package recipes adding the new setting to the conanfile ` settings =  "os", "compiler", "arch", "build_type", "libc"`

**Comment by theirix** [comment]
Thanks for the idea, @lasote.
I added settings, say `distr`, to `.conan/settings.yml` and default setting to `.conan/conan.conf`. 
It works but the main problem is that I need to modify all dependent packages and patch a `settings` line. Otherwise I get an error `'options.distr' doesn't exist`. If a version is updated it must be done again.

Is there a simple way to assume that all dependent packages have a custom settings value?

**Comment by memsharded** [comment]
Hi @theirix ,

I am not sure if I understood the problem. Maybe the error message is ``settings.distr` doesn't exist`` instead of ``options.distr doesn't exist``? Also, the "dependent packages", you mean, those packages that will be affected by ``distr``? Dependent packages is a common terminology for other meanings :) If you could please tell me a little more about this, I'll try to help.

Maybe an idea that could help: it is possible to define ``distr`` as a nested setting of ``os``, so the settings line of each package will remain unchanged, ``settings = "os", "compiler", ...

**Comment by theirix** [comment]
Hello, @memsharded, sorry for the delay. Finally found some time to test this.

In my previous comment I made an incorrect assumption about an error and messed with terminology. Of course I meant settings that could be set system-wide without modifying every conan package and ever without touching `conanfile.py`. Also 'Dependent packages' is 'requirements':)

Your idea about nested `os.distr` works perfectly.
So I replaced `os` definition in `settings.yml` with:
<pre>
os:
    Windows:
    Linux:
        distr: [None, el6, el7]
    Macos:
    Android:
    iOS:
</pre>

and added to `conan.conf` for every build machine:
<pre>
os=Linux
os.distr=None
</pre>

Now packages built on both `el6` and `el7` machines are distinct by `os.distr` setting. At other machines (i.e. osx) I can use the same `conanfile.py` and install binary packages just from conan.io.
